crates/rolldown/src/utils/renamer.rs

```diff
@@ -34,65 +27,149 @@ pub struct Renamer<'name> {
+      let name = self.get_unique_name(symbol_ref, &original_name);
+      self.canonical_names.insert(canonical_ref, name);
+    }
+  }
+
+  #[expect(clippy::map_entry)]
+  pub fn add_symbol_in_root_scope_with_original_name(
+    &mut self,
+    symbol_ref: SymbolRef,
+    original_name: &str,
+  ) {
+    let canonical_ref = symbol_ref.canonical_ref(self.symbol_db);
+
+    if !self.canonical_names.contains_key(&canonical_ref) {
+      let name = self.get_unique_name(symbol_ref, original_name);
+      self.canonical_names.insert(canonical_ref, name);
+    }
+  }
+
+  fn generate_candidate_name(original_name: &str, count: u32) -> Rstr {
+    concat_string!(original_name, "$", itoa::Buffer::new().format(count)).into()
+  }
+
+  fn get_module_used_names(
+    symbol_db: &'name SymbolRefDb,
+    canonical_ref: SymbolRef,
+  ) -> FxHashSet<&'name str> {
+    const RUNTIME_MODULE_INDEX: ModuleIdx = ModuleIdx::from_usize_unchecked(0);
+    if canonical_ref.owner.is_dummy() || canonical_ref.owner == RUNTIME_MODULE_INDEX {
+      FxHashSet::default()
+    } else {
+      let scoping = symbol_db.local_db(canonical_ref.owner).ast_scopes.scoping();
+      if scoping.symbols_len() == 0 {
+        return FxHashSet::default();
       }
-      Entry::Occupied(_) => {
-        // The symbol is already renamed
+      let root_symbol_ids =
+        scoping.get_bindings(scoping.root_scope_id()).values().collect::<FxHashSet<_>>();
+      scoping
+        .symbol_ids()
+        .zip(scoping.symbol_names())
+        .filter(|(symbol_id, _)| !root_symbol_ids.contains(symbol_id))
+        .map(|(_, name)| name)
+        .collect::<FxHashSet<&str>>()
+    }
+  }
+
+  fn get_unique_name(&mut self, canonical_ref: SymbolRef, original_name: &str) -> Rstr {
+    let (mut candidate_name, count) = match self.used_canonical_names.entry(original_name.to_rstr())
+    {
+      Entry::Occupied(o) => {
+        let count = o.into_mut();
+        *count += 1;
+        (Self::generate_candidate_name(original_name, *count), count)
       }
+      Entry::Vacant(v) => (original_name.to_rstr(), v.insert(0)),
+    };
+
+    loop {
+      let is_root_binding = self.entry_module.is_some_and(|module| {
+        let scoping = module.ast_scopes.scoping();
+        scoping.get_root_binding(&candidate_name).is_some_and(|symbol_id| {
+          let base_symbol = SymbolRef::from((module.owner_idx, symbol_id));
+          base_symbol == canonical_ref || base_symbol.canonical_ref(self.symbol_db) == canonical_ref
+        })
+      });
+
+      if (is_root_binding && !self.manual_reserved.contains(candidate_name.as_str()))
+        || !self.used_names.contains(&candidate_name)
+          && !(self.entry_module_used_names.contains(candidate_name.as_str()))
+          && !self
+            .module_used_names
+            .entry(canonical_ref.owner)
+            .or_insert_with(|| Self::get_module_used_names(self.symbol_db, canonical_ref))
+            .contains(candidate_name.as_str())
```

Comment:
How about merge the `manual_reserved /used_names/entry_module_used_names/module_used_names` into one `used_names`. It could reduce many hash.

Reply:
`manual_reserved` and `entry_module_used_names` has removed.

Unfortunately, `used_names` and `module_used_names` can't be merged into one due to their different uses.
`used_names` is a `FxHashSet<Rstr>` that is used to store names that are used for renaming.
`module_used_names` is a `FxHashSet<&name Rstr>` that stores the module used names, and is exactly used in what I said in https://github.com/rolldown/rolldown/pull/4014#issuecomment-2788185805.